# Importing Libraries & all dataset

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path=r'/Users/stefaniadaskalaki/Google Drive/Data Immersion/6 Achievement/CF6 BRecom'

## Importing customers dataset, performing consistency checks and descriptive statistics 

In [3]:
df_cust=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_customers_dataset.csv'),index_col=False)

In [4]:
#checking outcome

df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [5]:
df_cust.shape

(99441, 5)

In [6]:
df_cust.head(15)

customer_id                customer_unique_id  \
0   06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1   18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2   4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3   b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4   4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5   879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6   fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7   5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8   5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9   4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   
10  9fb35e4ed6f0a14a4977cd9aea4042bb  2a7745e1ed516b289ed9b29c7d0539a5   
11  5aa9e4fdd4dfd20959cad2d772509598  2a46fb94aef5cbeeb850418118cee090   
12  b2d1536598b73a9abd18e0d75d92f0a3  918dc87cd72cd9f6ed4bd442ed785235   
13  eabebad39a88bb6f5b52376faec28612  295c05e81917928d76245e842748184d   
14  1f1c7bf1c9b041b292af6c1c4470b753  3151a81801c8386361b62277d7fa5ecf   

    customer_zip_code_prefix          customer_city customer_state  
0                      14409                 franca             SP  
1                       9790  sao bernardo do campo             SP  
2                       1151              sao paulo             SP  
3                       8775        mogi das cruzes             SP  
4                      13056               campinas             SP  
5                      89254         jaragua do sul             SC  
6                       4534              sao paulo             SP  
7                      35182                timoteo             MG  
8                      81560               curitiba             PR  
9                      30575         belo horizonte             MG  
10                     39400          montes claros             MG  
11                     20231         rio de janeiro             RJ  
12                     18682       lencois paulista             SP  
13                      5704              sao paulo             SP  
14                     95110          caxias do sul             RS

### Checking for Missing Values

In [7]:
df_cust.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

### Checking for Columns Consistency

In [8]:
# checking for mixed type columns

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

### Checking for Duplicates

In [9]:
df_dups = df_cust[df_cust.duplicated()]

In [10]:
df_dups

Empty DataFrame
Columns: [customer_id, customer_unique_id, customer_zip_code_prefix, customer_city, customer_state]
Index: []

### Performing Descriptive Statistics

In [11]:
df_cust.describe()

customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

## Importing geolocation dataset, performing consistency checks and descriptive statistics 

In [12]:
df_geo=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_geolocation_dataset.csv'),index_col=False)

In [13]:
#checking outcome

df_geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [14]:
df_geo.shape

(1000163, 5)

I notice that total entries in the last dataset was 99441 so I will join them together with an inner join in order to make sure i have only those entries that will have complete information

In [15]:
df_geo.head(15)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                          1037       -23.545621       -46.639292   
1                          1046       -23.546081       -46.644820   
2                          1046       -23.546129       -46.642951   
3                          1041       -23.544392       -46.639499   
4                          1035       -23.541578       -46.641607   
5                          1012       -23.547762       -46.635361   
6                          1047       -23.546273       -46.641225   
7                          1013       -23.546923       -46.634264   
8                          1029       -23.543769       -46.634278   
9                          1011       -23.547640       -46.636032   
10                         1013       -23.547325       -46.634184   
11                         1032       -23.538418       -46.634778   
12                         1014       -23.546435       -46.633830   
13                         1012       -23.548946       -46.634671   
14                         1037       -23.545187       -46.637855   

   geolocation_city geolocation_state  
0         sao paulo                SP  
1         sao paulo                SP  
2         sao paulo                SP  
3         sao paulo                SP  
4         sao paulo                SP  
5         são paulo                SP  
6         sao paulo                SP  
7         sao paulo                SP  
8         sao paulo                SP  
9         sao paulo                SP  
10        sao paulo                SP  
11        sao paulo                SP  
12        sao paulo                SP  
13        sao paulo                SP  
14        são paulo                SP

### Checking for Missing Values

In [16]:
df_geo.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

No missing values

### Checking for Columns Consistency

In [17]:
# checking for mixed type columns

for col in df_geo.columns.tolist():
  weird = (df_geo[[col]].applymap(type) != df_geo[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_geo[weird]) > 0:
    print (col)

No mixed type columns

### Checking for Duplicates

In [18]:
df_dups_geo = df_geo[df_geo.duplicated()]

In [19]:
df_dups_geo

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
15                              1046       -23.546081       -46.644820   
44                              1046       -23.546081       -46.644820   
65                              1046       -23.546081       -46.644820   
66                              1009       -23.546935       -46.636588   
67                              1046       -23.546081       -46.644820   
...                              ...              ...              ...   
1000153                        99970       -28.343273       -51.873734   
1000154                        99950       -28.070493       -52.011342   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
15             sao paulo                SP  
44             sao paulo                SP  
65             sao paulo                SP  
66             sao paulo                SP  
67             sao paulo                SP  
...                  ...               ...  
1000153          ciriaco                RS  
1000154         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000162         tapejara                RS  

[261831 rows x 5 columns]

As we can see here there are a lot of dupplicates since we are talking about geolocation. There is no need to delete anything.

### Performing Descriptive Statistics

In [20]:
df_geo=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_geolocation_dataset.csv'),index_col=False)

In [21]:
df_geo.describe()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

## Importing order_item dataset, performing consistency checks and descriptive statistics 

In [22]:
df_ord_item=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_order_items_dataset.csv'),index_col=False)

In [23]:
#checking outcome

df_ord_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [24]:
df_ord_item.shape

(112650, 7)

Again I notice that this dataset has again more entries than the df_cust (99441) and df_geo(1000163)

In [25]:
df_ord_item.head(15)

order_id  order_item_id  \
0   00010242fe8c5a6d1ba2dd792cb16214              1   
1   00018f77f2f0320c557190d7a144bdd3              1   
2   000229ec398224ef6ca0657da4fc703e              1   
3   00024acbcdf0a6daa1e931b038114c75              1   
4   00042b26cf59d7ce69dfabb4e55b4fd9              1   
5   00048cc3ae777c65dbb7d2a0634bc1ea              1   
6   00054e8431b9d7675808bcb819fb4a32              1   
7   000576fe39319847cbb9d288c5617fa6              1   
8   0005a1a1728c9d785b8e2b08b904576c              1   
9   0005f50442cb953dcd1d21e1fb923495              1   
10  00061f2a7bc09da83e415a52dc8a4af1              1   
11  00063b381e2406b52ad429470734ebd5              1   
12  0006ec9db01a64e59a68b2c340bf65a7              1   
13  0008288aa423d2a3f00fcb17cd7d8719              1   
14  0008288aa423d2a3f00fcb17cd7d8719              2   

                          product_id                         seller_id  \
0   4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1   e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2   c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3   7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4   ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
5   ef92defde845ab8450f9d70c526ef70f  6426d21aca402a131fc0a5d0960a3c90   
6   8d4f2bb7e93e6710a28f34fa83ee7d28  7040e82f899a04d1b434b795a43b4617   
7   557d850972a7d6f792fd18ae1400d9b6  5996cddab893a4652a15592fb58ab8db   
8   310ae3c140ff94b03219ad0adc3c778f  a416b6a846a11724393025641d4edd5e   
9   4535b0e1091c278dfd193e5a1d63b39f  ba143b05f0110f0dc71ad71b4466ce92   
10  d63c1011f49d98b976c352955b1c4bea  cc419e0650a3c5ba77189a1882b7556a   
11  f177554ea93259a5b282f24e33f65ab6  8602a61d680a10a82cceeeda0d99ea3d   
12  99a4788cb24856965c36a24e339b6058  4a3ca9315b744ce9f8e9374361493884   
13  368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
14  368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   

    shipping_limit_date   price  freight_value  
0   2017-09-19 09:45:35   58.90          13.29  
1   2017-05-03 11:05:13  239.90          19.93  
2   2018-01-18 14:48:30  199.00          17.87  
3   2018-08-15 10:10:18   12.99          12.79  
4   2017-02-13 13:57:51  199.90          18.14  
5   2017-05-23 03:55:27   21.90          12.69  
6   2017-12-14 12:10:31   19.90          11.85  
7   2018-07-10 12:30:45  810.00          70.75  
8   2018-03-26 18:31:29  145.95          11.65  
9   2018-07-06 14:10:56   53.99          11.40  
10  2018-03-29 22:28:09   59.99           8.88  
11  2018-07-31 17:30:39   45.00          12.98  
12  2018-07-26 17:24:20   74.00          23.32  
13  2018-02-21 02:55:52   49.90          13.37  
14  2018-02-21 02:55:52   49.90          13.37

### Checking for Missing Values

In [26]:
df_ord_item.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

no missing values

### Checking for Consistency

In [27]:
# checking for mixed type columns

for col in df_ord_item.columns.tolist():
  weird = (df_ord_item[[col]].applymap(type) != df_ord_item[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord_item[weird]) > 0:
    print (col)

No mixed type columns

### Checking for Duplicates

In [28]:
df_dups_ord_item = df_ord_item[df_ord_item.duplicated()]

In [29]:
df_dups_ord_item

Empty DataFrame
Columns: [order_id, order_item_id, product_id, seller_id, shipping_limit_date, price, freight_value]
Index: []

No duplicates

### Performing Descriptive Statistics

In [30]:
df_ord_item.describe()

order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

### Change order_item_id data type

In [31]:
df_ord_item['order_item_id'] = df_ord_item['order_item_id'].astype(str)

In [32]:
# checking outcome

df_ord_item.describe()

price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000

## Importing olist_order_payments_dataset, performing consistency checks and descriptive statistics 

In [33]:
df_ord_pay=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_order_payments_dataset.csv'),index_col=False)

In [34]:
#checking outcome

df_ord_pay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [35]:
df_ord_pay.shape

(103886, 5)

In [36]:
df_ord_pay.head(15)

order_id  payment_sequential payment_type  \
0   b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1   a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2   25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3   ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4   42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
5   298fcdf1f73eb413e4d26d01b25bc1cd                   1  credit_card   
6   771ee386b001f06208a7419e4fc1bbd7                   1  credit_card   
7   3d7239c394a212faae122962df514ac7                   1  credit_card   
8   1f78449c87a54faf9e96e88ba1491fa9                   1  credit_card   
9   0573b5e23cbd798006520e1d5b4c6714                   1       boleto   
10  d88e0d5fa41661ce03cf6cf336527646                   1  credit_card   
11  2480f727e869fdeb397244a21b721b67                   1  credit_card   
12  616105c9352a9668c38303ad44e056cd                   1  credit_card   
13  cf95215a722f3ebf29e6bbab87a29e61                   1  credit_card   
14  769214176682788a92801d8907fa1b40                   1  credit_card   

    payment_installments  payment_value  
0                      8          99.33  
1                      1          24.39  
2                      1          65.71  
3                      8         107.78  
4                      2         128.45  
5                      2          96.12  
6                      1          81.16  
7                      3          51.84  
8                      6         341.09  
9                      1          51.95  
10                     8         188.73  
11                     1         141.90  
12                     1          75.78  
13                     5         102.66  
14                     4         105.28

### Checking for Missing Values

In [37]:
df_ord_pay.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

no missing values

### Checking for Consistency

In [38]:
# checking for mixed type columns

for col in df_ord_pay.columns.tolist():
  weird = (df_ord_pay[[col]].applymap(type) != df_ord_pay[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord_pay[weird]) > 0:
    print (col)

no mixed type columns

### Checking for Duplicates

In [39]:
df_ord_pay = df_ord_pay[df_ord_pay.duplicated()]

no duplicates

### Performing Descriptive Statistics

In [40]:
df_ord_pay.describe()

payment_sequential  payment_installments  payment_value
count                 0.0                   0.0            0.0
mean                  NaN                   NaN            NaN
std                   NaN                   NaN            NaN
min                   NaN                   NaN            NaN
25%                   NaN                   NaN            NaN
50%                   NaN                   NaN            NaN
75%                   NaN                   NaN            NaN
max                   NaN                   NaN            NaN

In [41]:
df_ord_pay['payment_sequential'].value_counts()

Series([], Name: payment_sequential, dtype: int64)

In [42]:
df_ord_pay['payment_sequential'].min()

nan

In [43]:
df_ord_pay['payment_sequential'].isnull().values.any()

False

In [44]:
check_for_nan = df_ord_pay.isnull()

In [45]:
print (check_for_nan)

Empty DataFrame
Columns: [order_id, payment_sequential, payment_type, payment_installments, payment_value]
Index: []


In [46]:
df_ord_pay['payment_sequential'].tolist()

[]

In [47]:
df_ord_pay['payment_sequential'] = df_ord_pay['payment_sequential'].astype(float)

In [48]:
df_ord_pay['payment_sequential'].min()

nan

In [49]:
df_ord_pay['payment_sequential'] = df_ord_pay['payment_sequential'].astype(int)

In [50]:
df_ord_pay['payment_sequential'].min()

nan

In [51]:
# importing the dataset again since I created an empty df_ord_pay when looking for duplicates

df_ord_pay=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_order_payments_dataset.csv'),index_col=False)

In [52]:
df_ord_pay.describe()

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

## Importing order_item dataset, performing consistency checks and descriptive statistics

In [3]:
df_ord=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_orders_dataset.csv'),index_col=False)

In [4]:
#checking outcome

df_ord.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [5]:
df_ord.shape

(99441, 8)

In [6]:
df_ord.head(15)

order_id                       customer_id  \
0   e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1   53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2   47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3   949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4   ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5   a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6   136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7   6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8   76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9   e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   
10  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
11  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
12  82566a660a982b15fb86e904c8d32918  d3e3b74c766bc6214e0c830b17ee2341   
13  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
14  432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   

   order_status order_purchase_timestamp    order_approved_at  \
0     delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1     delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2     delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3     delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4     delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5     delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
6      invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
7     delivered      2017-05-16 13:10:30  2017-05-16 13:22:11   
8     delivered      2017-01-23 18:29:09  2017-01-25 02:50:47   
9     delivered      2017-07-29 11:55:02  2017-07-29 12:05:32   
10    delivered      2017-05-16 19:41:10  2017-05-16 19:50:18   
11    delivered      2017-07-13 19:58:11  2017-07-13 20:10:08   
12    delivered      2018-06-07 10:06:19  2018-06-09 03:13:12   
13    delivered      2018-07-25 17:44:10  2018-07-25 17:55:14   
14    delivered      2018-03-01 14:14:28  2018-03-01 15:10:47   

   order_delivered_carrier_date order_delivered_customer_date  \
0           2017-10-04 19:55:00           2017-10-10 21:25:13   
1           2018-07-26 14:31:00           2018-08-07 15:27:45   
2           2018-08-08 13:50:00           2018-08-17 18:06:29   
3           2017-11-22 13:39:59           2017-12-02 00:28:42   
4           2018-02-14 19:46:34           2018-02-16 18:17:02   
5           2017-07-11 14:58:04           2017-07-26 10:57:55   
6                           NaN                           NaN   
7           2017-05-22 10:07:46           2017-05-26 12:55:51   
8           2017-01-26 14:16:31           2017-02-02 14:08:10   
9           2017-08-10 19:45:24           2017-08-16 17:14:30   
10          2017-05-18 11:40:40           2017-05-29 11:18:31   
11          2017-07-14 18:43:29           2017-07-19 14:04:48   
12          2018-06-11 13:29:00           2018-06-19 12:05:52   
13          2018-07-26 13:16:00           2018-07-30 15:52:25   
14          2018-03-02 21:09:20           2018-03-12 23:36:26   

   order_estimated_delivery_date  
0            2017-10-18 00:00:00  
1            2018-08-13 00:00:00  
2            2018-09-04 00:00:00  
3            2017-12-15 00:00:00  
4            2018-02-26 00:00:00  
5            2017-08-01 00:00:00  
6            2017-05-09 00:00:00  
7            2017-06-07 00:00:00  
8            2017-03-06 00:00:00  
9            2017-08-23 00:00:00  
10           2017-06-07 00:00:00  
11           2017-08-08 00:00:00  
12           2018-07-18 00:00:00  
13           2018-08-08 00:00:00  
14           2018-03-21 00:00:00

### Checking for Missing Values

In [7]:
df_ord.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [8]:
# dropping column order_delivered_customer_date 

df_ord.drop(columns = ['order_delivered_carrier_date','order_purchase_timestamp'])

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status    order_approved_at order_delivered_customer_date  \
0        delivered  2017-10-02 11:07:15           2017-10-10 21:25:13   
1        delivered  2018-07-26 03:24:27           2018-08-07 15:27:45   
2        delivered  2018-08-08 08:55:23           2018-08-17 18:06:29   
3        delivered  2017-11-18 19:45:59           2017-12-02 00:28:42   
4        delivered  2018-02-13 22:20:29           2018-02-16 18:17:02   
...            ...                  ...                           ...   
99436    delivered  2017-03-09 09:54:05           2017-03-17 15:08:01   
99437    delivered  2018-02-06 13:10:37           2018-02-28 17:37:56   
99438    delivered  2017-08-27 15:04:16           2017-09-21 11:24:17   
99439    delivered  2018-01-08 21:36:21           2018-01-25 23:32:54   
99440    delivered  2018-03-09 11:20:28           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0               2017-10-18 00:00:00  
1               2018-08-13 00:00:00  
2               2018-09-04 00:00:00  
3               2017-12-15 00:00:00  
4               2018-02-26 00:00:00  
...                             ...  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00  

[99441 rows x 6 columns]

In [9]:
# saving new df_ords without order_delivered_carrier_date, order_purchase_timestamp

df_ords=df_ord.drop(columns = ['order_delivered_carrier_date','order_purchase_timestamp'])

### Checking for Consistency

In [60]:
# checking for mixed type columns

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

order_approved_at
order_delivered_customer_date


In [61]:
# fixing mixed columns

df_ords['order_approved_at']= pd.to_datetime(df_ords['order_approved_at'])

In [62]:
df_ords['order_delivered_customer_date']= pd.to_datetime(df_ords['order_delivered_customer_date'])

### Checking for Duplicates

In [63]:
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, customer_id, order_status, order_approved_at, order_delivered_customer_date, order_estimated_delivery_date]
Index: []

no duplicates

### Performing Descriptive Statistics

In [64]:
df_ords.describe()

/var/folders/k8/6mb93rb95jd6bknf4qsf29_r0000gn/T/ipykernel_30091/2909690800.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_ords.describe()
/var/folders/k8/6mb93rb95jd6bknf4qsf29_r0000gn/T/ipykernel_30091/2909690800.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_ords.describe()


order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   
first                                NaN                               NaN   
last                                 NaN                               NaN   

       order_status    order_approved_at order_delivered_customer_date  \
count         99441                99281                         96476   
unique            8                90733                         95664   
top       delivered  2018-02-27 04:31:10           2018-05-08 23:38:46   
freq          96478                    9                             3   
first           NaN  2016-09-15 12:16:38           2016-10-11 13:46:32   
last            NaN  2018-09-03 17:40:06           2018-10-17 13:22:46   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522  
first                            NaN  
last                             NaN

## Importing products dataset, performing consistency checks and descriptive statistics

In [65]:
df_prod=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'olist_products_dataset.csv'),index_col=False)

In [66]:
#checking outcome

df_prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [67]:
df_prod.shape

(32951, 9)

In [68]:
df_prod.head(15)

product_id             product_category_name  \
0   1e9e8ef04dbcff4541ed26657ea517e5                        perfumaria   
1   3aa071139cb16b67ca9e5dea641aaa2f                             artes   
2   96bd76ec8810374ed1b65e291975717f                     esporte_lazer   
3   cef67bcfe19066a932b7673e239eb23d                             bebes   
4   9dc1a7de274444849c219cff195d0b71             utilidades_domesticas   
5   41d3672d4792049fa1779bb35283ed13             instrumentos_musicais   
6   732bd381ad09e530fe0a5f457d81becb                        cool_stuff   
7   2548af3e6e77a690cf3eb6368e9ab61e                  moveis_decoracao   
8   37cc742be07708b53a98702e77a21a02                  eletrodomesticos   
9   8c92109888e8cdf9d66dc7e463025574                        brinquedos   
10  14aa47b7fe5c25522b47b4b29c98dcb9                   cama_mesa_banho   
11  03b63c5fc16691530586ae020c345514                             bebes   
12  cf55509ea8edaaac1d28fdb16e48fc22             instrumentos_musicais   
13  7bb6f29c2be57716194f96496660c7c2                  moveis_decoracao   
14  eb31436580a610f202c859463d8c7415  construcao_ferramentas_seguranca   

    product_name_lenght  product_description_lenght  product_photos_qty  \
0                  40.0                       287.0                 1.0   
1                  44.0                       276.0                 1.0   
2                  46.0                       250.0                 1.0   
3                  27.0                       261.0                 1.0   
4                  37.0                       402.0                 4.0   
5                  60.0                       745.0                 1.0   
6                  56.0                      1272.0                 4.0   
7                  56.0                       184.0                 2.0   
8                  57.0                       163.0                 1.0   
9                  36.0                      1156.0                 1.0   
10                 54.0                       630.0                 1.0   
11                 49.0                       728.0                 4.0   
12                 43.0                      1827.0                 3.0   
13                 51.0                      2083.0                 2.0   
14                 59.0                      1602.0                 4.0   

    product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0              225.0               16.0               10.0              14.0  
1             1000.0               30.0               18.0              20.0  
2              154.0               18.0                9.0              15.0  
3              371.0               26.0                4.0              26.0  
4              625.0               20.0               17.0              13.0  
5              200.0               38.0                5.0              11.0  
6            18350.0               70.0               24.0              44.0  
7              900.0               40.0                8.0              40.0  
8              400.0               27.0               13.0              17.0  
9              600.0               17.0               10.0              12.0  
10            1100.0               16.0               10.0              16.0  
11            7150.0               50.0               19.0              45.0  
12             250.0               17.0                7.0              17.0  
13             600.0               68.0               11.0              13.0  
14             200.0               17.0                7.0              17.0

### Checking for Missing Values

In [69]:
df_prod.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [70]:
# deleting all values with NaN since it way below 5%

df_prod_dup = df_prod.dropna()

In [71]:
# checking outcome

df_prod_dup.isna().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

### Checking for Consistency

In [72]:
# checking for mixed type columns

for col in df_prod_dup.columns.tolist():
  weird = (df_prod_dup[[col]].applymap(type) != df_prod_dup[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prod_dup[weird]) > 0:
    print (col)

no mixed columns

### Checking for Duplicates

In [73]:
df_prod_dup[df_prod_dup.duplicated()]

Empty DataFrame
Columns: [product_id, product_category_name, product_name_lenght, product_description_lenght, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm]
Index: []

no duplicates found

### Performing Descriptive Statistics

In [74]:
df_prod_dup.describe()

product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.636115   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32340.000000  
mean          23.208596  
std           12.078762  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

## Importing product_category_name_translation dataset, performing consistency checks and descriptive statistics

In [75]:
df_prod_cat=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'product_category_name_translation.csv'),index_col=False)

In [76]:
#checking outcome

df_prod_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [77]:
df_prod_cat.shape

(71, 2)

In [78]:
df_prod_cat.head(15)

product_category_name product_category_name_english
0               beleza_saude                 health_beauty
1     informatica_acessorios         computers_accessories
2                 automotivo                          auto
3            cama_mesa_banho                bed_bath_table
4           moveis_decoracao               furniture_decor
5              esporte_lazer                sports_leisure
6                 perfumaria                     perfumery
7      utilidades_domesticas                    housewares
8                  telefonia                     telephony
9         relogios_presentes                 watches_gifts
10         alimentos_bebidas                    food_drink
11                     bebes                          baby
12                 papelaria                    stationery
13  tablets_impressao_imagem        tablets_printing_image
14                brinquedos                          toys

### Checking for Missing Values

In [79]:
df_prod_cat.isna().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

### Checking for Consistency

In [80]:
# checking for mixed type columns

for col in df_prod_cat.columns.tolist():
  weird = (df_prod_cat[[col]].applymap(type) != df_prod_cat[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prod_cat[weird]) > 0:
    print (col)

### Checking for Duplicates

In [81]:
df_prod_cat[df_prod_cat.duplicated()]

Empty DataFrame
Columns: [product_category_name, product_category_name_english]
Index: []

## Exporting all dfs to csv files